aws-datastore/src/main/java/com/amplifyframework/datastore/DataStoreConfiguration.java

```diff
@@ -0,0 +1,340 @@
+
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.util.Iterator;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A user-provided configuration for the DataStore.
+ */
+@SuppressWarnings("unused")
+public final class DataStoreConfiguration {
+    static final String PLUGIN_CONFIG_KEY = "awsDataStorePlugin";
+    static final long DEFAULT_SYNC_INTERVAL_MINUTES = TimeUnit.DAYS.toMinutes(1);
+    static final int DEFAULT_SYNC_MAX_RECORDS = 1_000;
+    static final int DEFAULT_SYNC_PAGE_SIZE = 1_000;
+
+    private static final Logger LOG = Amplify.Logging.forNamespace("amplify:aws-datastore");
+
+    private final DataStoreErrorHandler dataStoreErrorHandler;
+    private final DataStoreConflictHandler dataStoreConflictHandler;
+    private final long syncIntervalInMinutes;
+    private final int syncMaxRecords;
+    private final int syncPageSize;
+
+    private DataStoreConfiguration(
+            DataStoreErrorHandler dataStoreErrorHandler,
+            DataStoreConflictHandler dataStoreConflictHandler,
+            long syncIntervalInMinutes,
+            int syncMaxRecords,
+            int syncPageSize) {
+        this.dataStoreErrorHandler = dataStoreErrorHandler;
+        this.dataStoreConflictHandler = dataStoreConflictHandler;
+        this.syncIntervalInMinutes = syncIntervalInMinutes;
+        this.syncMaxRecords = syncMaxRecords;
+        this.syncPageSize = syncPageSize;
+    }
+
+    /**
+     * Begin building a new instance of {@link DataStoreConfiguration}.
+     * @return A new builder instance
+     */
+    @NonNull
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * Begin building a new instance of {@link DataStoreConfiguration} by reading DataStore
+     * settings from the config file.
+     * @param pluginJson DataStore plugin configuration from amplicationconfiguration.json
+     * @param userProvidedConfiguration An instance of {@DataStoreConfiguration} with settings specified by the user
+     *                                  which will be used as overrides.
+     * @return A new builder instance
+     * @throws DataStoreException exception thrown if there's an unexpected configuration key or
+     * an invalid configuration value
+     */
+    @NonNull
+    public static Builder builder(@NonNull JSONObject pluginJson,
+                                  @NonNull DataStoreConfiguration userProvidedConfiguration) throws DataStoreException {
+        return builder(pluginJson)
+            .syncPageSize(userProvidedConfiguration.getSyncPageSize())
+            .syncMaxRecords(userProvidedConfiguration.getSyncMaxRecords())
+            .syncIntervalMs(userProvidedConfiguration.getSyncIntervalInMinutes())
+            .dataStoreErrorHandler(userProvidedConfiguration.getDataStoreErrorHandler())
+            .dataStoreConflictHandler(userProvidedConfiguration.getDataStoreConflictHandler());
```

Comment:
Hm, how does this  work? Doesn't the user config always clobber the file config?

Reply:
> non-default values provided in userProvidedConfiguration to override whatever was in the JSON

I think this means that you'll still overwrite config file values, if they aren't the defaults. No?

I like the second mechanism you propose, better, I think.

Basically:
1. Create a config from the file json. This will have 0...all values as null;
2. Create a config from user. This will have 0...all values as null;
3. Merge the two configs according to the rule:
```
if (userProvided.getSomeConfigValue() != null) {
    use(userProvided.getSomeConfigValue());
} else {
    use(configFileProvided.getSomeConfigValue());
```
4. As a final step, evaluate the merged config. If any fields are still `null`, then use a default value.

That is kind of a lot of machinery isn't it (eye-roll) sigh.

Reply:
> I think this means that you'll still overwrite config file values, if they aren't the defaults. No?

Correct. The assumption here is that any non-default values passed via the `userProvidedConfiguration` will clobber anything that was set via the file. That being said, if we want to respect any non-default settings set via the config file, I can add an extra check in the setters:

```
public Builder syncMaxRecords(@IntRange(from = 0) int syncMaxRecords) {
    //Only set this value if the current value is default AND the incoming value is NOT default
    if (this.syncMaxRecords == DEFAULT_SYNC_MAX_RECORDS && 
        syncMaxRecords != DEFAULT_SYNC_MAX_RECORDS) {
        this.syncMaxRecords = syncMaxRecords;
    }
    return Builder.this;
}
```

I think in the end, it comes down to deciding which is more authoritative: the config file or the java object passed in via the constructor.

The drawback of making them nullable is that you need to have that extra step to ensure defaults are set. When I started going that route, there was no clear place for that. I couldn't do it on the build method of the builder, because I actually need the values to remain null. So I would need something (probably in the plugin's configure method) that would set those defaults.

Another thing I just thought of if we want to go the null route, is that I could modify the getters to return their respective defaults if the underlying field is null. I may give that a try.

Reply:
That actually worked, so I updated the code to make the fields null (with the exception of the handler) to start with and then I only write to them if the underlying field is null. The getters, in turn, return a default value if the underlying field is null

Reply:
Changes made in the [latest commit](https://github.com/aws-amplify/amplify-android/pull/401/commits/6c3c4c5522dcd21c055d511c03afb970e64bede5)

- Removed the logic from the getters and modified the builder's build method to ensure defaults when instructed to do so.